Specifying just one of -s, -c yields TypeScript compilation errors

Each writes code that requires the other's dependencies.

# `-c fetch` with no `-s` fails
```
flutterby:interfaces ariels$ ./node_modules/.bin/concord -o ./gen node -c fetch foo@0.0.1 v2.ts
npm WARN foo@0.0.1 No license field.

removed 74 packages and audited 14 packages in 1.106s
found 0 vulnerabilities

src/koaMW.ts:2:28 - error TS2307: Cannot find module 'koa-router'.

2 import * as koaRouter from 'koa-router';
                             ~~~~~~~~~~~~

src/koaMW.ts:3:26 - error TS2307: Cannot find module 'koa-bodyparser'.

3 import * as koaBody from 'koa-bodyparser'; // tslint:disable-line:no-unused-variable
                           ~~~~~~~~~~~~~~~~

src/server.ts:3:22 - error TS2307: Cannot find module 'koa'.

3 import * as Koa from 'koa';
                       ~~~~~

src/server.ts:4:25 - error TS2307: Cannot find module 'koa-router'.

4 import * as Router from 'koa-router';
                          ~~~~~~~~~~~~

src/server.ts:5:29 - error TS2307: Cannot find module 'koa-bodyparser'.

5 import * as bodyParser from 'koa-bodyparser';
                              ~~~~~~~~~~~~~~~~

src/server.ts:41:54 - error TS7006: Parameter 'ctx' implicitly has an 'any' type.

41     this.koaRouter.post('/reportConcurrency', async (ctx) => {
                                                        ~~~


Found 6 errors.

Failed to generate files:
Error: command: 'tsc' failed process failed with exit code: 2
    at ChildProcess.child.on (/Users/ariels/Google Drive/Work/Dev/xenu/interfaces/node_modules/concord/dist/utils.js:12:20)
    at ChildProcess.emit (events.js:197:13)
    at maybeClose (internal/child_process.js:984:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:265:5)
```

# `-s koa` with no `-c` fails
```
flutterby:interfaces ariels$ ./node_modules/.bin/concord -o ./gen node -s koa foo@0.0.1 v2.ts
npm WARN foo@0.0.1 No license field.

added 74 packages from 67 contributors, removed 4 packages and audited 174 packages in 2.741s
found 0 vulnerabilities

src/client.ts:2:19 - error TS2307: Cannot find module 'node-fetch'.

2 import fetch from 'node-fetch';
                    ~~~~~~~~~~~~

src/client.ts:3:29 - error TS2307: Cannot find module 'node-fetch'.

3 import { RequestInit } from 'node-fetch';
                              ~~~~~~~~~~~~

src/client.ts:4:29 - error TS2307: Cannot find module 'abort-controller'.

4 import AbortController from 'abort-controller';
                              ~~~~~~~~~~~~~~~~~~

src/client.ts:47:60 - error TS2739: Type '{}' is missing the following properties from type 'Options': follow, agent, redirect, compress

47   public constructor(protected readonly serverUrl: string, protected readonly options: Options = {}) {
                                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 4 errors.

Failed to generate files:
Error: command: 'tsc' failed process failed with exit code: 2
    at ChildProcess.child.on (/Users/ariels/Google Drive/Work/Dev/xenu/interfaces/node_modules/concord/dist/utils.js:12:20)
    at ChildProcess.emit (events.js:197:13)
    at maybeClose (internal/child_process.js:984:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:265:5)
```

--- a/README.md
+++ b/README.md
@@ -27,7 +27,7 @@ The heavy lifting is done by [typescript-json-schema](https://github.com/YousefE
     ```
 1. Compile the schema.
     ```bash
-    mkdir -p ./generated && concord -o ./generated node --client fetch --server koa example@0.0.1 interface.ts
+    concord -o ./generated node --client fetch --server koa example@0.0.1 interface.ts
     ```
 1. Write the server code.
 
@@ -252,14 +252,14 @@ Defining `returns: null | SomethingElse` on a method will compile to `Promise<nu
 ### Binaris backend
 1. Create a new directory.
 1. Switch to new directory.
-1. Run `bn create node8 add`.
 1. Create an `interface.ts` file as shown above.
-1. Generate your server code with `concord node --client fetch --server binaris example@0.0.1 interface.ts -o .` (the client is here for the sake of the example and is not required).
+1. Generate your server code with `concord node --client fetch --server binaris example@0.0.1 interface.ts -o generated` (the client is here for the sake of the example and is not required).
+1. Run `bn create node8 add`.
 1. Implement a handler:
 
     *`function.js`*
     ```js
-    const { ExampleWrapper } = require('./server');
+    const { ExampleWrapper } = require('./generated/server');
     exports.handler = ExampleWrapper.add(async (a, b) => a + b);
     ```
 1. Export your binaris credentials: `export BINARIS_ACCOUNT_ID=$(bn show accountId) BINARIS_API_KEY=$(bn show apiKey)`
@@ -268,7 +268,7 @@ Defining `returns: null | SomethingElse` on a method will compile to `Promise<nu
 
     *`test.js`*
     ```js
-    const { ExampleClient } = require('./client');
+    const { ExampleClient } = require('./generated/client');
 
     async function main() {
       const url = `https://run.binaris.com/v2/run/${process.env.BINARIS_ACCOUNT_ID}`;

--- a/src/cli.ts
+++ b/src/cli.ts
@@ -124,7 +124,7 @@ async function main({
 
   const genPath = output || mktemp();
   try {
-    const generator = await TSOutput.create(genPath);
+    const generator = new TSOutput(genPath);
     const generated = await generate(runtime, paths, { client, server });
     await generator.write(runtime, name, version, generated, { client, server });
     if (!noCompile) {

--- a/src/cli.ts
+++ b/src/cli.ts
@@ -1,4 +1,5 @@
-import { format } from 'util';
+import { format, promisify } from 'util';
+import * as glob from 'glob';
 import { randomBytes } from 'crypto';
 import { mkdir } from 'mz/fs';
 import * as path from 'path';
@@ -117,13 +118,15 @@ async function main({
       throw new Error('Must specify exactly one of (client or server) option with `publish`');
     }
   }
-  const genPath = output || mktemp();
-  if (genPath !== output) {
-    await mkdir(genPath);
+  const paths = await promisify(glob)(pattern);
+  if (paths.length === 0) {
+    throw new Error(`No paths found with pattern: ${pattern}`);
   }
+
+  const genPath = output || mktemp();
   try {
     const generator = await TSOutput.create(genPath);
-    const generated = await generate(runtime, pattern, { client, server });
+    const generated = await generate(runtime, paths, { client, server });
     await generator.write(runtime, name, version, generated, { client, server });
     if (!noCompile) {
       await generator.compile();

--- a/src/cli.ts
+++ b/src/cli.ts
@@ -1,7 +1,6 @@
 import { format, promisify } from 'util';
 import * as glob from 'glob';
 import { randomBytes } from 'crypto';
-import { mkdir } from 'mz/fs';
 import * as path from 'path';
 import * as yargs from 'yargs';
 import * as rmrf from 'rmfr';

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,10 +1,8 @@
 import { readFile } from 'mz/fs';
 import { isArray, zip, fromPairs, merge, mergeWith } from 'lodash';
-import * as glob from 'glob';
 import * as ts from 'typescript';
 import * as tjs from 'typescript-json-schema';
 import * as mustache from 'mustache';
-import { promisify } from 'util';
 import * as path from 'path';
 import { GeneratedCode, Package, FrameworkMap, Runtime } from './types';
 import { transform } from './transform';
@@ -105,10 +103,9 @@ function getGenerator(runtime: Runtime, kind: string, framework: string): Genera
 
 export async function generate(
   runtime: Runtime,
-  filePattern: string,
+  paths: string[],
   frameworks: FrameworkMap
 ): Promise<GeneratedCode> {
-  const paths = await promisify(glob)(filePattern);
   const settings: tjs.PartialArgs = {
     required: true,
     noExtraProps: true,

--- a/src/output.ts
+++ b/src/output.ts
@@ -1,6 +1,6 @@
 import * as path from 'path';
 import { merge } from 'lodash';
-import { mkdir, exists, writeFile } from 'mz/fs';
+import { mkdir, writeFile } from 'mz/fs';
 import { GeneratedCode, FrameworkMap, Runtime } from './types';
 import { spawn } from './utils';
 
@@ -40,17 +40,9 @@ export class TSOutput {
   protected npm: (...args: string[]) => Promise<number>;
   protected tsc: (...args: string[]) => Promise<number>;
 
-  protected constructor(protected readonly genPath: string) {
+  constructor(protected readonly genPath: string) {
     this.npm = (...args: string[]) => spawn('npm', args, { cwd: genPath, stdio: 'inherit' });
-    this.tsc = (...args: string[]) => spawn('tsc', args, { cwd: this.genPath, stdio: 'inherit' });
-  }
-
-  public static async create(genPath) {
-    const dirExists = await exists(genPath);
-    if (dirExists) {
-      throw new Error(`output dir: ${genPath} exists`);
-    }
-    return new this(genPath);
+    this.tsc = (...args: string[]) => spawn('tsc', args, { cwd: genPath, stdio: 'inherit' });
   }
 
   public async write(

--- a/src/output.ts
+++ b/src/output.ts
@@ -1,6 +1,6 @@
 import * as path from 'path';
 import { merge } from 'lodash';
-import { mkdir, stat, writeFile } from 'mz/fs';
+import { mkdir, stat, exists, writeFile } from 'mz/fs';
 import { GeneratedCode, FrameworkMap, Runtime } from './types';
 import { spawn } from './utils';
 
@@ -46,9 +46,9 @@ export class TSOutput {
   }
 
   public static async create(genPath) {
-    const st = await stat(genPath);
-    if (!st.isDirectory()) {
-      throw new Error(`output dir: ${genPath} is not a directory`);
+    const dirExists = await exists(genPath);
+    if (dirExists) {
+      throw new Error(`output dir: ${genPath} exists`);
     }
     return new this(genPath);
   }
@@ -60,6 +60,8 @@ export class TSOutput {
     { code, pkg: basePackage }: GeneratedCode,
     { client, server }: FrameworkMap
   ) {
+    await mkdir(this.genPath);
+
     try {
       await mkdir(path.join(this.genPath, 'src'));
     } catch (err) {

--- a/src/output.ts
+++ b/src/output.ts
@@ -1,6 +1,6 @@
 import * as path from 'path';
 import { merge } from 'lodash';
-import { mkdir, stat, exists, writeFile } from 'mz/fs';
+import { mkdir, exists, writeFile } from 'mz/fs';
 import { GeneratedCode, FrameworkMap, Runtime } from './types';
 import { spawn } from './utils';
 

--- a/src/test/test_binaris.ts
+++ b/src/test/test_binaris.ts
@@ -1,7 +1,6 @@
 import test from 'ava';
 import * as path from 'path';
 import { writeFile } from 'mz/fs';
-import { spawn } from '../utils';
 import { pass, TestRunner } from './utils';
 
 class TestCase extends TestRunner {

--- a/src/test/test_binaris.ts
+++ b/src/test/test_binaris.ts
@@ -67,11 +67,11 @@ main().catch((err) => {
       '--server', 'binaris',
       '--nocompile',
       '-o',
-      '.',
+      'gen',
     );
 
-    await writeFile(path.join(this.dir, 'src', 'run.ts'), this.runner);
-    await writeFile(path.join(this.dir, 'binaris.yml'), JSON.stringify({
+    await writeFile(path.join(this.genDir, 'src', 'run.ts'), this.runner);
+    await writeFile(path.join(this.genDir, 'binaris.yml'), JSON.stringify({
       functions: {
         [this.functionName]: {
           file: 'handler.js',
@@ -81,8 +81,8 @@ main().catch((err) => {
         },
       },
     }));
-    await writeFile(path.join(this.dir, 'src', 'handler.ts'), this.handler);
-    await writeFile(path.join(this.dir, 'src', 'test.ts'), `
+    await writeFile(path.join(this.genDir, 'src', 'handler.ts'), this.handler);
+    await writeFile(path.join(this.genDir, 'src', 'test.ts'), `
 import { expect, use } from 'chai';
 import * as chaiAsPromised from 'chai-as-promised';
 
@@ -98,9 +98,9 @@ ${this.tester}`);
       '@types/chai-as-promised',
     );
 
-    await this.spawn('npm', 'install');
-    await this.spawn(this.inModules('tsc'));
-    await this.spawn(this.inModules('bn'), 'deploy', this.functionName);
+    await this.spawnInGen('npm', 'install');
+    await this.spawnInGen(this.inModules('tsc'));
+    await this.spawnInGen(this.inModules('bn'), 'deploy', this.functionName);
   }
 
   public async cleanup(): Promise<void> {
@@ -109,11 +109,7 @@ ${this.tester}`);
   }
 
   public async exec(): Promise<void> {
-    await spawn(
-      'node', ['run.js'], {
-      cwd: this.dir,
-      stdio: 'inherit',
-    });
+    await this.spawnInGen('node', 'run.js');
   }
 }
 

--- a/src/test/test_browser.ts
+++ b/src/test/test_browser.ts
@@ -2,7 +2,7 @@ import test from 'ava';
 import * as path from 'path';
 import { launch } from 'puppeteer';
 import { spawn as origSpawn } from 'child_process';
-import { writeFile, mkdir } from 'mz/fs';
+import { writeFile } from 'mz/fs';
 import { spawn } from '../utils';
 import { pass, TestRunner } from './utils';
 

--- a/src/test/test_browser.ts
+++ b/src/test/test_browser.ts
@@ -62,12 +62,7 @@ main().catch((err) => {
   }
 
   public async setup() {
-    await mkdir(path.join(this.dir, 'server'));
-    await mkdir(path.join(this.dir, 'client'));
-    await mkdir(path.join(this.dir, 'server', 'src'));
     await writeFile(path.join(this.dir, 'schema.ts'), this.schema);
-    await writeFile(path.join(this.dir, 'server', 'src', 'main.ts'), this.main);
-    await writeFile(path.join(this.dir, 'server', 'src', 'handler.ts'), this.handler);
     await spawn('cp', [
       '-r',
       path.join(__dirname, '..', '..', 'webpack-test'),
@@ -98,6 +93,9 @@ main().catch((err) => {
       cwd: path.join(this.dir),
       stdio: 'inherit',
     });
+
+    await writeFile(path.join(this.dir, 'server', 'src', 'main.ts'), this.main);
+    await writeFile(path.join(this.dir, 'server', 'src', 'handler.ts'), this.handler);
 
     await spawn('npm', [
       'install',

--- a/src/test/test_generate.ts
+++ b/src/test/test_generate.ts
@@ -30,7 +30,6 @@ class TestCase {
   public async generate(frameworks: FrameworkMap): Promise<GeneratedCode> {
     await mkdir(this.dir);
     try {
-      await mkdir(path.join(this.dir, 'gen'));
       const schemaPath = path.join(this.dir, 'schema.ts');
       await writeFile(schemaPath, this.schema);
       await spawn('node', [

--- a/src/test/test_rpc.ts
+++ b/src/test/test_rpc.ts
@@ -48,18 +48,8 @@ main().catch((err) => {
 
   public async setup() {
     await mkdir(this.dir);
-    await mkdir(path.join(this.dir, 'src'));
-    const schemaPath = path.join(this.dir, 'schema.ts');
-    await writeFile(schemaPath, this.schema);
-    await writeFile(path.join(this.dir, 'src', 'main.ts'), this.main);
-    await writeFile(path.join(this.dir, 'src', 'handler.ts'), this.handler);
-    await writeFile(path.join(this.dir, 'src', 'test.ts'), `
-import { expect, use } from 'chai';
-import * as chaiAsPromised from 'chai-as-promised';
-
-use(chaiAsPromised);
-${this.tester}`);
-
+    const genDir = path.join(this.dir, 'gen');
+    await writeFile(path.join(this.dir, 'schema.ts'), this.schema);
     await spawn('node', [
       path.join(__dirname, '..', 'cli.js'),
       'node',
@@ -69,11 +59,20 @@ ${this.tester}`);
       '--server', 'koa',
       '--nocompile',
       '-o',
-      '.',
+      'gen',
     ], {
       cwd: this.dir,
       stdio: 'inherit',
     });
+
+    await writeFile(path.join(genDir, 'src', 'main.ts'), this.main);
+    await writeFile(path.join(genDir, 'src', 'handler.ts'), this.handler);
+    await writeFile(path.join(genDir, 'src', 'test.ts'), `
+import { expect, use } from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+
+use(chaiAsPromised);
+${this.tester}`);
 
     await spawn('npm', [
       'install',
@@ -82,19 +81,19 @@ ${this.tester}`);
       '@types/chai',
       '@types/chai-as-promised',
     ], {
-      cwd: this.dir,
+      cwd: genDir,
       stdio: 'inherit',
     });
 
     await spawn('npm', ['install'], {
-      cwd: this.dir,
+      cwd: genDir,
       stdio: 'inherit',
     });
 
     await spawn(
       path.join(__dirname, '..', '..', 'node_modules', '.bin', 'tsc'),
       [], {
-      cwd: this.dir,
+      cwd: genDir,
       stdio: 'inherit',
     });
   }
@@ -104,7 +103,7 @@ ${this.tester}`);
   }
 
   public async exec(): Promise<{ stdout: string, stderr: string }> {
-    const testPath = path.join(this.dir, 'main.js');
+    const testPath = path.join(this.dir, 'gen', 'main.js');
     const [stdout, stderr] = await exec(`node ${testPath}`);
     return { stdout: stdout.toString(), stderr: stderr.toString() };
   }

--- a/src/test/utils.ts
+++ b/src/test/utils.ts
@@ -22,6 +22,10 @@ export class TestRunner {
     return mkdtemp(path.join(tmpdir(), 'concord-test-'), { encoding: 'utf8' });
   }
 
+  public get genDir(): string {
+    return path.join(this.dir, 'gen');
+  }
+
   public inModules(exec: string): string {
     return path.join(__dirname, '..', '..', 'node_modules', '.bin', exec);
   }
@@ -29,6 +33,13 @@ export class TestRunner {
   public spawn(cmd, ...args) {
     return spawn(cmd, args, {
       cwd: this.dir,
+      stdio: 'inherit',
+    });
+  }
+
+  public spawnInGen(cmd, ...args) {
+    return spawn(cmd, args, {
+      cwd: this.genDir,
       stdio: 'inherit',
     });
   }

